oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/reactive/function/client/ServerOAuth2AuthorizedClientExchangeFilterFunction.java

```diff
@@ -101,6 +143,7 @@
 	public ServerOAuth2AuthorizedClientExchangeFilterFunction(ReactiveOAuth2AuthorizedClientManager authorizedClientManager) {
 		Assert.notNull(authorizedClientManager, "authorizedClientManager cannot be null");
 		this.authorizedClientManager = authorizedClientManager;
+		this.responseHandler = createDefaultResponseHandler(authorizedClientManager);
```

Comment:
I'm not sure this makes sense here. If the application supplies `ReactiveOAuth2AuthorizedClientManager` then it's responsible for configuring all it's collaborator's. I think we only need to set the default handlers when constructing the default manager in `createDefaultAuthorizedClientManager()`.

I see that you were trying to reuse the manager handlers in `createDefaultResponseHandler()`. Given that I commented on removing the getters for the handlers, this won't be possible. As mentioned in another comment, the implementation is quite minimal so code reuse wouldn't make much of a difference here. I'm ok with minimal code duplication vs. exposing a public API. Easier from a maintenance standpoint on our end and allows us to change things more easily.

Reply:
I'm trying to make the basic functionality of removing cached clients on HTTP 401 from a resource server easy to use out of the box.

If all the default success/failure handler classes (which provide the basic functionality) are made private, and the getters are removed from the client managers, then it's going to be fairly painful for an application that uses the `ServerOAuth2AuthorizedClientExchangeFilterFunction(ReactiveOAuth2AuthorizedClientManager authorizedClientManager)` constructor to remove cached clients.

Right now, the `ServerOAuth2AuthorizedClientExchangeFilterFunction`'s response handler (which delegates to the client manager's failure handler to perform cached client removal) is private.  We would need to make a public interface for this.  This is not terrible, but I was trying to keep this all private.

An app that wanted to use the `ServerOAuth2AuthorizedClientExchangeFilterFunction(ReactiveOAuth2AuthorizedClientManager authorizedClientManager)` constructor, and remove cached clients on HTTP 401 responses from the resource server would need to:
* Implement a `ReactiveOAuth2AuthorizationFailureHandler` that is a copy/paste of `AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.DefaultAuthorizationFailureHandler` or `DefaultReactiveOAuth2AuthorizedClientManager.DefaultAuthorizationFailureHandler` 
* Create the `ReactiveOAuth2AuthorizedClientManager` instance, and set its failure handler to the custom implementation
* Implement a copy/paste of `ServerOAuth2AuthorizedClientExchangeFilterFunction.DefaultResponseHandler` that delegates to the `ReactiveOAuth2AuthorizationFailureHandler`
* Create the `ServerOAuth2AuthorizedClientExchangeFilterFunction` instance, and set its response handler

All this is very painful, especially for just enabling basic functionality.

How do you suggest making the removal of cached clients on HTTP 401 from a resource server easy to use by an app that uses the `ServerOAuth2AuthorizedClientExchangeFilterFunction(ReactiveOAuth2AuthorizedClientManager authorizedClientManager)` constructor?  (maybe I'm missing something)

Reply:
@philsttr 

I've been giving this some further thought and I agree that we should make it easier for users to leverage an existing basic (default) implementation.

So let's say we provide 2 implementations: `SaveAuthorizedClientReactiveOAuth2AuthorizationSuccessHandler` and `RemoveAuthorizedClientReactiveOAuth2AuthorizationFailureHandler`.

Both would be public outer classes and therefore could be used by any requiring component. For example, `DefaultReactiveOAuth2AuthorizedClientManager` and `AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager` would instantiate both as their default handlers. A setter would be exposed for each handler but no getter. We could apply the same changes to `ServerOAuth2AuthorizedClientExchangeFilterFunction` as well.

For `SaveAuthorizedClientReactiveOAuth2AuthorizationSuccessHandler`, we could provide 2 constructors - one that accepts `ServerOAuth2AuthorizedClientRepository` and the other that accepts `ReactiveOAuth2AuthorizedClientService`.

What are your thoughts?

Reply:
I've updated the PR with these changes.

I still don't quite like that the `ServerOAuth2AuthorizedClientExchangeFilterFunction(ReactiveOAuth2AuthorizedClientManager authorizedClientManager)` doesn't automatically handle removal of the authorized client on auth failure, without additional configuration of calling `setAuthorizationFailureHandler`, but I can live with this, given the constraints of keeping the getters private.

I've document this limitation in the javadocs.